src/actions.rs

```diff
@@ -0,0 +1,621 @@
+#[cfg(feature = "mysql")]
+pub(crate) async fn save_policy<'a>(
+    conn: &ConnectionPool,
+    rules: Vec<NewCasbinRule<'a>>,
+) -> Result<()> {
+    let mut transaction = conn
+        .begin()
+        .await
+        .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+    sqlx::query!("DELETE FROM casbin_rules")
+        .execute(&mut transaction)
+        .await
+        .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+    for rule in rules {
+        sqlx::query!(
+            "INSERT INTO casbin_rules ( ptype, v0, v1, v2, v3, v4, v5 )
+                 VALUES ( ?, ?, ?, ?, ?, ?, ? )",
+            rule.ptype,
+            rule.v0,
+            rule.v1,
+            rule.v2,
+            rule.v3,
+            rule.v4,
+            rule.v5
+        )
+        .execute(&mut transaction)
+        .await
+        .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+    }
+    transaction
+        .commit()
+        .await
+        .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+    Ok(())
+}
+
+#[cfg(feature = "postgres")]
+pub(crate) async fn add_policy<'a>(
+    mut conn: &ConnectionPool,
+    rule: NewCasbinRule<'a>,
+) -> Result<bool> {
+    sqlx::query!(
+        "INSERT INTO casbin_rules ( ptype, v0, v1, v2, v3, v4, v5 )
+                 VALUES ( $1, $2, $3, $4, $5, $6, $7 )",
+        rule.ptype,
+        rule.v0,
+        rule.v1,
+        rule.v2,
+        rule.v3,
+        rule.v4,
+        rule.v5
+    )
+    .execute(&mut conn)
+    .await
+    .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+
+    Ok(true)
+}
+
+#[cfg(feature = "mysql")]
+pub(crate) async fn add_policy<'a>(
+    mut conn: &ConnectionPool,
+    rule: NewCasbinRule<'a>,
+) -> Result<bool> {
+    sqlx::query!(
+        "INSERT INTO casbin_rules ( ptype, v0, v1, v2, v3, v4, v5 )
+                 VALUES ( ?, ?, ?, ?, ?, ?, ? )",
+        rule.ptype,
+        rule.v0,
+        rule.v1,
+        rule.v2,
+        rule.v3,
+        rule.v4,
+        rule.v5
+    )
+    .execute(&mut conn)
+    .await
+    .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+
+    Ok(true)
+}
+
+#[cfg(feature = "postgres")]
+pub(crate) async fn add_policies<'a>(
+    conn: &ConnectionPool,
+    rules: Vec<NewCasbinRule<'a>>,
+) -> Result<bool> {
+    let mut transaction = conn
+        .begin()
+        .await
+        .map_err(|err| CasbinError::from(AdapterError(Box::new(Error::SqlxError(err)))))?;
+    for rule in rules {
+        sqlx::query!(
+            "INSERT INTO casbin_rules ( ptype, v0, v1, v2, v3, v4, v5 )
+                 VALUES ( $1, $2, $3, $4, $5, $6, $7 )",
+            rule.ptype,
+            rule.v0,
+            rule.v1,
+            rule.v2,
+            rule.v3,
+            rule.v4,
+            rule.v5
+        )
+        .execute(&mut transaction)
+        .await
```

Comment:
we need to check if it has been added, `if n != 1` then we stop the loop and revert the transaction